File: Backtracking/SubsetsII.java
/*
Given a collection of integers that might contain duplicates, nums, return all possible subsets.

Note: The solution set must not contain duplicate subsets.

For example,
If nums = [1,2,2], a solution is:

[
  [2],
  [1],
  [1,2,2],
  [2,2],
  [1,2],
  []
]
*/

public class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        return bt(nums, new ArrayList<List<Integer>>(), new ArrayList<Integer>(), 0);
    }
    private List<List<Integer>> bt(int[] nums, List<List<Integer>> list, List<Integer> path, int start){
        list.add(new ArrayList<Integer>(path));
        for(int i = start; i < nums.length; i++){
            if(i > start && nums[i] == nums[i - 1]) continue;
            path.add(nums[i]);
            bt(nums, list, path, i + 1);
            path.remove(path.size() - 1);
        }
        return list;
    }
}

File: Backtracking/SubsetsII.js
/*
Given a collection of integers that might contain duplicates, nums, return all possible subsets.

Note: The solution set must not contain duplicate subsets.

For example,
If nums = [1,2,2], a solution is:

[
  [2],
  [1],
  [1,2,2],
  [2,2],
  [1,2],
  []
]
*/

/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsetsWithDup = function(nums) {
    var bt = function(nums, path, start, res){
        res.push(path.slice());
        for(var i = start; i < nums.length; i++){
            if(i != start && nums[i - 1] == nums[i]) continue;
            path.push(nums[i]);
            bt(nums, path, i + 1, res);
            path.pop();
        }
        return res;
    }
    nums.sort();
    return bt(nums, [], 0, []);
};